src/main/java/com/metamx/common/CompressionUtils.java

```diff
@@ -0,0 +1,388 @@
+
+public class CompressionUtils
+{
+  private static final Logger log = new Logger(CompressionUtils.class);
+  private static final int DEFAULT_RETRY_COUNT = 3;
+
+  public static final String GZ_SUFFIX = ".gz";
+  public static final String ZIP_SUFFIX = ".zip";
+
+  public static long zip(File directory, File outputZipFile) throws IOException
+  {
+    if (!isZip(outputZipFile.getName())) {
+      log.warn("No .zip suffix[%s], putting files from [%s] into it anyway.", outputZipFile, directory);
+    }
+
+    try (final FileOutputStream out = new FileOutputStream(outputZipFile)) {
+      return zip(directory, out);
+    }
+  }
+
+  public static long zip(File directory, OutputStream out) throws IOException
+  {
+    if (!directory.isDirectory()) {
+      throw new IOException(String.format("directory[%s] is not a directory", directory));
+    }
+    final File[] files = directory.listFiles();
+
+    long totalSize = 0;
+    try (final ZipOutputStream zipOut = new ZipOutputStream(out)) {
+      for (File file : files) {
+        log.info("Adding file[%s] with size[%,d].  Total size so far[%,d]", file, file.length(), totalSize);
+        if (file.length() >= Integer.MAX_VALUE) {
+          zipOut.finish();
+          throw new IOException(String.format("file[%s] too large [%,d]", file, file.length()));
+        }
+        zipOut.putNextEntry(new ZipEntry(file.getName()));
+        totalSize += Files.asByteSource(file).copyTo(zipOut);
+      }
+      zipOut.closeEntry();
+    }
+
+    return totalSize;
+  }
+
+  public static FileUtils.FileCopyResult unzip(
+      final ByteSource byteSource,
+      final File outDir,
+      final Predicate<Throwable> shouldRetry,
+      boolean cacheLocally
+  ) throws IOException
+  {
+    if (!cacheLocally) {
+      try {
+        return RetryUtils.retry(
+            new Callable<FileUtils.FileCopyResult>()
+            {
+              @Override
+              public FileUtils.FileCopyResult call() throws Exception
+              {
+                return unzip(byteSource.openStream(), outDir);
+              }
+            },
+            shouldRetry,
+            DEFAULT_RETRY_COUNT
+        );
+      }
+      catch (Exception e) {
+        throw Throwables.propagate(e);
+      }
+    }
+    final File tmpFile = File.createTempFile("compressionUtilZipCache", ZIP_SUFFIX);
+    try {
+      FileUtils.FileCopyResult copyResult = FileUtils.retryCopy(
+          byteSource,
+          tmpFile,
+          shouldRetry,
+          DEFAULT_RETRY_COUNT
+      );
+      return unzip(tmpFile, outDir);
+    }
+    finally {
+      if (!tmpFile.delete()) {
+        log.warn("Could not delete zip cache at [%s]", tmpFile.toString());
+      }
+    }
+  }
+
+  public static FileUtils.FileCopyResult unzip(
+      final ByteSource byteSource,
+      final File outDir,
+      boolean cacheLocally
+  ) throws IOException
+  {
+    return unzip(byteSource, outDir, Predicates.<Throwable>alwaysTrue(), cacheLocally);
```

Comment:
I think that by default we should retry on Exceptions but not on other kinds of Throwables.